homework/homework5/node.py

```diff
@@ -0,0 +1,173 @@
+from copy import deepcopy
+from random import random
+from typing import Tuple, Optional, Any
+from json import dumps, JSONEncoder, JSONDecoder
+
+
+class Node:
+    priority: float
+
+    left_child: Optional["Node"] = None
+    right_child: Optional["Node"] = None
+
+    def __init__(
+        self,
+        key,
+        value,
+        priority: Optional[float] = None,
+        left_child: Optional["Node"] = None,
+        right_child: Optional["Node"] = None,
+    ):
+        self.key = key
+        self.value = value
+
+        if priority is None:
+            self.priority = random()
+        else:
+            self.priority = priority
+
+        self.left_child = left_child
+        self.right_child = right_child
+
+    def __str__(self):
+        return dumps(self, allow_nan=True, cls=self.NodeEncoder)
+
+    def __eq__(self, other):
+        if isinstance(other, Node):
+            return (
+                self.key == other.key
+                and self.value == other.value
+                and self.priority == other.priority
+                and self.left_child == other.left_child
+                and self.right_child == other.right_child
+            )
+        return False
+
+    def __contains__(self, key) -> bool:
+        if self.key == key:
+            return True
+
+        if (self.left_child is not None) and (key < self.key):
+            return key in self.left_child
+
+        if (self.right_child is not None) and (key > self.key):
+            return key in self.right_child
+
+        return False
+
+    def __iter__(self):
+        yield self.value
+        if self.left_child is not None:
+            yield from self.left_child
+        if self.right_child is not None:
+            yield from self.right_child
+
+    def __reversed__(self):
+        if self.left_child is not None:
+            yield from reversed(self.left_child)
+        if self.right_child is not None:
+            yield from reversed(self.right_child)
+        yield self.value
+
+    def insert(self, key, value, priority: Optional[float] = None) -> "Node":
+        new_node = Node(key, value, priority)
+
+        left_split, right_split = self.__split(key)
+
+        if left_split is None:
+            return new_node.__merge(right_split)
+        else:
+            left_split = left_split.__merge(new_node)
+            return left_split.__merge(right_split)
+
+    def get(self, key) -> Any:
+        if self.key == key:
+            return self.value
+
+        if (self.left_child is not None) and (key < self.key):
```

Comment:
You often use this condition and the simular with `right_child`. Probably you can create a function for it, what do you think?

Reply:
I put checking these conditions into separate functions, but because of problems with mypy (it can't understand that checking for None is already done) I had to write assert.